src/cairo.rs

```diff
@@ -0,0 +1,247 @@
+) -> Result<CairoRunResult, anyhow::Error> {
+    match args.cairo_version {
+        CairoVersion::cairo0 => run_cairo0(args, tmp_dir).map_err(|e| anyhow::anyhow!(e)),
+        CairoVersion::cairo1 => run_cairo1(args, tmp_dir),
+    }
+}
+
+/// Runs a Cairo 0 program and generates the necessary outputs for proving
+///
+/// # Arguments
+///
+/// * `prove_args` - The arguments for the prove command
+/// * `tmp_dir` - A temporary directory to store intermediate files
+///
+/// # Returns
+///
+/// A `Result` containing `CairoRunResult` on success, or an `Error` on failure
+///
+/// # Errors
+///
+/// This function can return various errors related to file I/O, program execution,
+/// trace encoding, and public input generation.
+pub fn run_cairo0(
+    prove_args: &ProveArgs,
+    tmp_dir: &tempfile::TempDir,
+) -> Result<CairoRunResult, Error> {
+    let filename = prove_args
+        .cairo_program
+        .file_stem()
+        .unwrap()
+        .to_str()
+        .unwrap();
+
+    let program = Program::from_file(&prove_args.cairo_program, Some("main"))?;
+
+    let mut hint_processor = BuiltinHintProcessor::new_empty();
+
+    let cairo_run_config = CairoRunConfig {
+        entrypoint: "main",
+        trace_enabled: true,
+        relocate_mem: true,
+        layout: get_layout(&prove_args.layout),
+        proof_mode: true,
+        secure_run: None,
+        disable_trace_padding: false,
+        allow_missing_builtins: None,
+    };
+
+    let runner = cairo_run_program(&program, &cairo_run_config, &mut hint_processor)?;
+    let relocated_trace = runner
+        .relocated_trace
+        .as_ref()
+        .ok_or(Error::Trace(TraceError::TraceNotRelocated))?;
+
+    let trace_path = tmp_dir.path().join(format!("{}_trace.json", filename));
+    let trace_file = std::fs::File::create(trace_path.clone())?;
+    let mut trace_writer =
+        FileWriter::new(io::BufWriter::with_capacity(3 * 1024 * 1024, trace_file));
+    write_encoded_trace(relocated_trace, &mut trace_writer)?;
+    trace_writer.flush()?;
+
+    let memory_path = tmp_dir.path().join(format!("{}_memory.json", filename));
+    let memory_file = std::fs::File::create(memory_path.clone())?;
+    let mut memory_writer =
+        FileWriter::new(io::BufWriter::with_capacity(5 * 1024 * 1024, memory_file));
+    write_encoded_memory(&runner.relocated_memory, &mut memory_writer)?;
+    memory_writer.flush()?;
+
+    let air_public_input_path = tmp_dir
+        .path()
+        .join(format!("{}_air_public_input.json", filename));
+    let air_public_input = runner.get_air_public_input()?.serialize_json()?;
+    std::fs::write(air_public_input_path.clone(), air_public_input)?;
+
+    let air_private_input_path = tmp_dir
+        .path()
+        .join(format!("{}_air_private_input.json", filename));
+    let trace_absolute_path = trace_path
+        .as_path()
+        .canonicalize()
+        .unwrap_or(trace_path.clone())
+        .to_string_lossy()
+        .to_string();
```

Comment:
Why not use [.to_str](https://doc.rust-lang.org/stable/std/path/struct.PathBuf.html#method.to_str)? Do we want it to be lossy or fail?

Reply:
Unintended change. Fixed.